helix-vcs/src/differ.rs

```diff
@@ -0,0 +1,164 @@
+use std::mem::take;
+use std::ops::{Deref, Range};
+use std::sync::Arc;
+
+use arc_swap::ArcSwap;
+use helix_core::{Rope, RopeSlice};
+use imara_diff::intern::InternedInput;
+use imara_diff::Algorithm;
+use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
+use tokio::task::JoinHandle;
+use tokio::time::{timeout, Duration};
+
+use crate::rope_line_cache::InternedRopeLines;
+use crate::{LineDiff, LineDiffs};
+
+#[cfg(test)]
+mod test;
+
+#[derive(Clone, Debug)]
+pub struct Differ {
+    channel: UnboundedSender<Event>,
+    line_diffs: Arc<ArcSwap<LineDiffs>>,
+}
+
+impl Differ {
+    pub fn new(diff_base: Rope, doc: Rope) -> Differ {
+        Differ::new_with_handle(diff_base, doc).0
+    }
+
+    fn new_with_handle(diff_base: Rope, doc: Rope) -> (Differ, JoinHandle<()>) {
+        let (sender, receiver) = unbounded_channel();
+        let line_diffs: Arc<ArcSwap<LineDiffs>> = Arc::default();
+        let worker = DiffWorker {
+            channel: receiver,
+            line_diffs: line_diffs.clone(),
+            new_line_diffs: LineDiffs::default(),
+        };
+        let handle = tokio::spawn(worker.run(diff_base, doc));
+        let differ = Differ {
+            channel: sender,
+            line_diffs,
+        };
+        (differ, handle)
+    }
+    pub fn get_line_diffs(&self) -> impl Deref<Target = impl Deref<Target = LineDiffs>> {
+        self.line_diffs.load()
+    }
+
+    pub fn update_document(&self, doc: Rope) -> bool {
+        self.channel.send(Event::UpdateDocument(doc)).is_ok()
+    }
+
+    pub fn update_diff_base(&self, diff_base: Rope) -> bool {
+        self.channel.send(Event::UpdateDiffBase(diff_base)).is_ok()
+    }
+}
+
+// TODO configuration
+const DIFF_MAX_DEBOUNCE: u32 = 10;
+const DIFF_DEBOUNCE_TIME: u64 = 1;
+const MAX_DIFF_LEN: usize = u16::MAX as usize;
+const ALGORITHM: Algorithm = Algorithm::Histogram;
+
+struct DiffWorker {
+    channel: UnboundedReceiver<Event>,
+    line_diffs: Arc<ArcSwap<LineDiffs>>,
+    new_line_diffs: LineDiffs,
+}
+
+impl DiffWorker {
+    async fn run(mut self, diff_base: Rope, doc: Rope) {
+        let mut interner = InternedRopeLines::new(diff_base, doc);
+        self.perform_diff(interner.interned_lines());
+        self.apply_line_diff();
+        while let Some(event) = self.channel.recv().await {
+            let mut accumulator = EventAccumulator::new();
+            accumulator.handle_event(event);
+            accumulator
+                .accumulate_debounced_events(&mut self.channel)
+                .await;
+
+            if let Some(new_doc) = accumulator.doc {
+                interner.update_doc(new_doc)
+            }
+            if let Some(new_base) = accumulator.diff_base {
+                interner.update_diff_base(new_base)
+            }
+
+            self.perform_diff(interner.interned_lines());
+            self.apply_line_diff();
```

Comment:
Don't we need an event to notify differ than diff is done?

Reply:
I actually figured out a way to do this without major changes after all thanks to @archseer suggestion in the matrix chat :tada: